query/src/main/java/org/infinispan/query/impl/IndexedQueryImpl.java

```diff
@@ -158,14 +163,19 @@ public QueryResult<?> execute() {
          }
 
          long start = queryStatistics.isEnabled() ? System.nanoTime() : 0;
-         SearchQuery<E> searchQuery = (SearchQuery<E>) searchQueryBuilder.build();
-         SearchResult<E> searchResult = searchQuery.fetch(queryDefinition.getFirstResult(), queryDefinition.getMaxResults());
+         SearchQuery<?> searchQuery = searchQueryBuilder.build();
+         SearchResult<?> searchResult = searchQuery.fetch(queryDefinition.getFirstResult(), queryDefinition.getMaxResults());
          if (queryStatistics.isEnabled()) recordQuery(System.nanoTime() - start);
 
          return new QueryResultImpl<>(
                // the hit count cannot exceed the cache size
                new TotalHitCount((int) searchResult.total().hitCountLowerBound(), searchResult.total().isHitCountExact()),
-               searchResult.hits());
+               searchResult.hits().stream().map(origin -> {
+                  if (origin instanceof EntityLoaded) {
```

Comment:
wondering if the Lucene query can return `EntityLoaded` all the time with `null` metadata when it is not required. Just to avoid the `instanceof` conditions.

Reply:
good point! let's test it though

Reply:
no in case of aggregations different objects are returned...

Reply:
This method can use the new instanceof assignment as well to avoid the cast.

Reply:
good idea! Doing it